compiler/frontend/src/ast_to_hir.rs

```diff
@@ -143,41 +143,42 @@ struct Context<'a> {
 }
 
 impl Context<'_> {
-    #[must_use]
-    fn start_non_top_level(&mut self) -> NonTopLevelResetState {
-        NonTopLevelResetState(mem::replace(&mut self.is_top_level, false))
-    }
-    #[allow(clippy::needless_pass_by_value)]
-    fn end_non_top_level(&mut self, reset_state: NonTopLevelResetState) {
-        self.is_top_level = reset_state.0;
+    fn with_non_top_level<F, T>(&mut self, func: F) -> T
+    where
+        F: Fn(&mut Self) -> T,
+    {
+        let reset_state = mem::replace(&mut self.is_top_level, false);
+        let res = func(self);
+        self.is_top_level = reset_state;
+        res
     }
-}
-struct NonTopLevelResetState(bool);
 
-impl Context<'_> {
     #[must_use]
-    fn start_scope(&mut self) -> ScopeResetState {
-        ScopeResetState {
+    fn with_scope<F, T>(&mut self, id_prefix: Option<hir::Id>, func: F) -> (Body, T)
```

Comment:
You can change `id_prefix` to accept an `impl Into<Option<hir::Id>>`, then callers can omit the `Some(…)` around that value